### ISSUE TITLE
[SD-XL Img2Img] Batched img2img inputs cause "fishnet" artifacts and visible generative watermarks

### ISSUE THREAD

**Issue by bghira** [huggingface/diffusers#4297]
### Describe the bug

I have been scratching my head over this for days, as suddenly the quality of my outputs dropped off a cliff. In efforts to nail down the issue,

* Updated pytorch to nightly, no change
* Updated xformers to git master, no change
* Updated triton, pretty much every other dependency
* Used git main branch for Diffusers
* Disabled/uninstalled Compel 2.0.0
* Reverted Diffusers to 0.18.x without mixture-of-experts code
* Switched samplers, step size, strength.

None of these helped, other than reducing the strength of the refiner.

However, reducing the batch size to 1 reproduces the results from the Diffusers example code.

Moreover, there just seems to be general issues with the handling of prompt sizes, generator lists, and string/list prompts. If the prompt is a list, the negative prompt must be non-empty, etc.

# Batch size 1
![1690351189 3764806602677297301877020f68e1cd1797410](https://github.com/huggingface/diffusers/assets/59658056/c956ad59-afd1-43dc-8a0a-f5132a4dbd9a)
# Batch size 2
![1690323387 3962464b0d71d0cf9e2aadbd6819fdb6b92364d](https://github.com/huggingface/diffusers/assets/59658056/9510de44-6832-4920-b90b-577b4c5899f3)

### Reproduction

```py
from diffusers import DiffusionPipeline
import torch
torch.backends.cudnn.deterministic = False
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
torch.backends.cudnn.benchmark = True
torch.backends.cuda.enable_flash_sdp(False)

base_pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-0.9", torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
base_pipe.to("cuda") # OR, pipe.enable_sequential_cpu_offload() OR,
#pipe.enable_model_cpu_offload()

# if using torch < 2.0
# pipe.enable_xformers_memory_efficient_attention()

# Reproducibility.
torch_seed = 4202420420
refiner_seed = 698008569
refiner_strength = 0.50
prompt = "happy child flying a kite on a sunny day"
negative_prompt = ''
# Batch size.
batch_size = 4
prompt = [ prompt ] * batch_size
negative_prompt = [ negative_prompt ] * batch_size
# We're going to schedule 20 steps, and complete 50% of them using either model.
total_num_steps = 5
# We need multiple Generators.
generator = [ torch.Generator(device="cuda").manual_seed(torch_seed) ] * batch_size

pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-refiner-0.9", torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
# Using channels last layout.
pipe.unet.to(memory_format=torch.channels_last)
pipe.to("cuda") # OR, pipe.enable_sequential_cpu_offload() OR, 

# Generate the base image.
pre_image = base_pipe(prompt=prompt, output_type="pil", generator=generator,
        num_inference_steps=total_num_steps, negative_prompt=negative_prompt).images

# Generate a range from 0.1 to 0.9, with 0.1 increments.
test_strengths = [x / 10 for x in range(1, 10)]
for refiner_strength in test_strengths:
    # Generate a new set of random states for each image.
    generator_two = [ torch.Generator(device="cuda").manual_seed(refiner_seed) ] * batch_size
    # Put through the refiner now.
    images = pipe(prompt=prompt, image=pre_image, aesthetic_score=10, negative_aesthetic_score=2.4, generator=generator_two,
                num_inference_steps=total_num_steps, strength=refiner_strength, negative_prompt=negative_prompt).images # denoising_start
    for idx in range(0, len(images)):
        print(f'Image: {idx}')
        images[idx].save(f'test-{refiner_strength}-{idx}.png', format='PNG')
```

### System Info

- `diffusers` version: 0.19.0.dev0
- Platform: Linux-5.19.0-45-generic-x86_64-with-glibc2.31
- Python version: 3.9.16
- PyTorch version (GPU?): 2.1.0.dev20230725+cu118 (True)
- Huggingface_hub version: 0.16.4
- Transformers version: 4.30.2
- Accelerate version: 0.18.0
- xFormers version: not installed
- Using GPU in script?: A100 80G
- Using distributed or parallel set-up in script?: False

### Who can help?

@patrickvonplaten

**Comment by patrickvonplaten** [comment]
Thanks for the nice repro - checking!

**Comment by patrickvonplaten** [comment]
BTW does it only happen for the refiner or also for pure "base"?

**Comment by patrickvonplaten** [comment]
Ah I figured what the problem is, it's the watermarker. Very nice issue here, let's try to make this an informative guide for the community. 

So what happened here is that we start seeing weird artificats with the hypotheses that it's because of the batch size.
I'm challenging this assumption now a bit saying it's not because of the batch size **but because the base model does not return "latents", but "pil" instead**.

What happens here in my opinion is the following. In the above script run the following pipeline:

```
latents -> unet_base -> vae_decode -> watermark -> image_base -> vae_encode -> unet_refiner -> vae_decode -> watermark
```

What is the problems?! We're encoding a watermarked image and denoise it again and decode it again. **We should not do this**. Let's make sure that we always keep the output in latents if we plan on running a second pipeline just after.

1.) I've adapted your script slightly for easier testing and I've set the strength of the refiner only to 0.2 as recommended by the authors.

The script then looks as follows.

```py
#!/usr/bin/env python3
from diffusers import DiffusionPipeline
import torch
torch.backends.cudnn.deterministic = False
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
torch.backends.cudnn.benchmark = True
torch.backends.cuda.enable_flash_sdp(False)

base_pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-0.9", torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
base_pipe.to("cuda") # OR, pipe.enable_sequential_cpu_offload() OR,

# Reproducibility.
torch_seed = 4202420420
refiner_seed = 698008569
refiner_strength = 0.50
prompt = "happy child flying a kite on a sunny day"
negative_prompt = ''
# Batch size.
batch_size = 2
do_latent = False
prompt = [ prompt ] * batch_size
negative_prompt = [ negative_prompt ] * batch_size
# We're going to schedule 20 steps, and complete 50% of them using either model.
total_num_steps = 20
# We need multiple Generators.
generator = [ torch.Generator(device="cuda").manual_seed(torch_seed) ] * batch_size

pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-refiner-0.9", vae=base_pipe.vae, text_encoder_2=base_pipe.text_encoder_2, torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
# Using channels last layout.
pipe.unet.to(memory_format=torch.channels_last)
pipe.to("cuda") # OR, pipe.enable_sequential_cpu_offload() OR, 

# Generate the base image.
pre_image = base_pipe(prompt=prompt, generator=generator,
        num_inference_steps=total_num_steps, negative_prompt=negative_prompt, output_type="latent" if do_latent else "pil").images

# Generate a range from 0.1 to 0.9, with 0.1 increments.
test_strengths = [0.2]
for refiner_strength in test_strengths:
    # Generate a new set of random states for each image.
    generator_two = [ torch.Generator(device="cuda").manual_seed(refiner_seed) ] * batch_size
    # Put through the refiner now.
    images = pipe(prompt=prompt, image=pre_image, aesthetic_score=10, negative_aesthetic_score=2.4, generator=generator_two,
                num_inference_steps=total_num_steps, strength=refiner_strength, negative_prompt=negative_prompt).images # denoising_start
    for idx in range(0, len(images)):
        print(f'Image: {idx}')
        images[idx].save(f'/home/patrick/images/refiner_bug/test-{refiner_strength}-{idx}--{batch_size}--{do_latent}.png', format='PNG')
```

Ok now let's run the script four times, and display the results below.

**Note, generally it's almost impossible to get 1-to-1 the same results even with all the seeding because different batch sizes use different kernels**. Also see [this](https://github.com/huggingface/diffusers/issues/2659) issue.


1.) batch_size = 1, do_latents = False
![test-0 2-0--1--False](https://github.com/huggingface/diffusers/assets/23423619/25579ed6-b60e-4116-8e73-1f65dc682df4)

2.) batch_size = 1, do_latents = True
![test-0 2-0--1--True](https://github.com/huggingface/diffusers/assets/23423619/3c62b986-08f7-4787-adbc-0afef631ea90)

3.) batch_size = 2, do_latents = False
![test-0 2-0--2--False](https://github.com/huggingface/diffusers/assets/23423619/03ce7696-8ccb-404a-9c67-dbf212cdeefd)

4.) batch_size = 2, do_latents = True
![test-0 2-0--2--True](https://github.com/huggingface/diffusers/assets/23423619/d1479a6d-b311-4fe6-b443-1ef83a1373b3)

Note how you can see the artifact only for 1.) and 3.), which means also for batch size = 1! 

The reason seems to be clearly because of the watermarker. Hence, let's make sure to **not** output "pil" or "pt" images for the base model if we're planning on later refining it or else we need to otherwise skip the watermarker and only add it in the final stage.

**Comment by patrickvonplaten** [comment]
I've changed the title here a bit to clarify for future readers what's going on. Let me know if this makes sense to you @bghira

**Comment by bghira** [comment]
hi Patrick, thank you for the investigation. this occurs with pil or latent results. my code in production where we initially observed this, uses mixture-of-experts diffusion with latents.

however, when i reverted my change where batches are processed as a batch, the issue went away. I have to loop over each image separately, which means the issue should still be happening, but it is not.

**Comment by bghira** [comment]
@patrickvonplaten i also wanted to point out that when using batch_size 1 i can even use the refiner with up to 100% of the process without this noise appearing.

when the batch size is > 1, i can not do this. around 40-50% of the process being run through the refiner will then result in the "fishnet" watermark artifacts.

when batch size is 1, i can run the refiner at 10, 20, 30, 40, 50, 60, 70, 80, 90%+ strengths, and it will work correctly.

**Comment by patrickvonplaten** [comment]
Hmm ok that's interesting! I was not able to reproduce this with the code above sadly, as you can see the no watermark batched image looks fine. Any chance you could maybe reproduce the problem in a google colab that we can both work on or a different code snippet here?

**Comment by bghira** [comment]
in your code, once I extend the refiner to 50% of the process, this pattern takes over. even with a batch size of 1.

in ComfyUI, this does not happen, even at 90% of the refiner. as well, if I'm doing single batches, I have a lot more success.

it seems the `happy child` prompt might be skewing things toward watermark, after searching LAION a bit... all of the happy children are watermarked! it feels weird, but maybe putting `happy child` into negative prompt is best :D

![image](https://github.com/huggingface/diffusers/assets/59658056/e4ca0ac4-c4e0-41ed-b125-a36bbca23d19)

**Comment by bghira** [comment]
![image](https://github.com/huggingface/diffusers/assets/59658056/9675f349-19da-46c8-a460-39eb91eabcda)
^ after the fix I applied,

![image](https://github.com/huggingface/diffusers/assets/59658056/dcbd0a6b-3c86-484b-8d1f-d952a36fe78d)
^ before

![image](https://github.com/huggingface/diffusers/assets/59658056/2dd2793a-504f-4154-8f94-e62b1b99b93d)
![image](https://github.com/huggingface/diffusers/assets/59658056/50f57223-abf9-4afb-9aed-710ba7cd2b62)

For the same settings otherwise, there is a stunning loss in quality with batch sizes.

**Comment by patrickvonplaten** [comment]
and this happens only with 0.9?

**Comment by bghira** [comment]
1.0 is worse.
![image](https://github.com/huggingface/diffusers/assets/59658056/e54f439c-84cf-4d8a-82cb-c640acbb2161)

**Comment by bghira** [comment]
batch size 1
![image](https://github.com/huggingface/diffusers/assets/59658056/86110780-d599-46ee-8134-92ea7215bcdd)

**Comment by patrickvonplaten** [comment]
> batch size 1 at 50% refiner strength ![image](https://user-images.githubusercontent.com/59658056/256670520-86110780-d599-46ee-8134-92ea7215bcdd.png)

What is the problem with this picture? 

Also I recommend using [this](https://huggingface.co/stabilityai/sdxl-vae) VAE and why the strength so high? The refiner has only been trained on strength up to 0.2.

**Comment by patrickvonplaten** [comment]
I'm getting reasonable pictures: 
![test-0 2-0--1--True](https://github.com/huggingface/diffusers/assets/23423619/e8b0603a-e7cc-4e7e-900a-93e33eda69c3)

**Comment by bghira** [comment]
sorry, that image has no issue - it's the correct output we expect, when using batch size 1. but when using batch size 2, we get the results earlier.

**Comment by patrickvonplaten** [comment]
Ah gotcha ok, let me try again with batch size 2

**Comment by bghira** [comment]
i think it might stem from something specific to the pipeline's handling of `num_images_per_prompt` as that seemingly always needs to be `1` for the img2img pipeline. in order to increase the batch size, i have had to do something i traditionally haven't, which is give a list of duplicate prompts - and then i have to provide multiple Generators, and it's just very unusual behaviour.

**Comment by patrickvonplaten** [comment]
Hmm, I can't reproduce any differences between batch size 1 and 2 (we also have tests that check exactly this).

Does the error happen only when using `num_images_per_prompt` because you don't use it here: https://github.com/huggingface/diffusers/issues/4297#issue-1823053849

**Comment by bghira** [comment]
a different refiner seed didn't seem to make the issue go away like changing the batch size does. this is another test of the same 1.0 model at 0.5 with two prompts.

![image](https://github.com/huggingface/diffusers/assets/59658056/b28034ba-a406-4689-bea3-647657ce0c8b)
![image](https://github.com/huggingface/diffusers/assets/59658056/72ecc324-7598-4e12-867c-b876de43b2eb)

**Comment by patrickvonplaten** [comment]
Can you send me the code you use?

**Comment by bghira** [comment]
i have not used `num_images_per_prompt` in any of my tests, it simply hasn't worked that way for me.

**Comment by bghira** [comment]
certainly. i have just modified it now to have `denoising_end`/`denoising_begin` specified instead of strength to see if the issue goes away, but it does not.

```py
from diffusers import DiffusionPipeline
import torch
torch.backends.cudnn.deterministic = False
torch.backends.cuda.matmul.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = False
torch.backends.cudnn.benchmark = True
torch.backends.cuda.enable_flash_sdp(False)

base_pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
base_pipe.to("cuda") # OR, pipe.enable_sequential_cpu_offload() OR,

# Reproducibility.
torch_seed = 4202420420
refiner_seed = 698008569
refiner_strength = 0.50
prompt = "happy child flying a kite on a sunny day"
negative_prompt = ''
# Batch size.
batch_size = 2
do_latent = True

# We're going to schedule 20 steps, and complete 50% of them using either model.
total_num_steps = 20
# We need multiple Generators.
generator = torch.Generator(device="cuda").manual_seed(torch_seed)

pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-xl-refiner-1.0", vae=base_pipe.vae, text_encoder_2=base_pipe.text_encoder_2, torch_dtype=torch.float16, use_safetensors=True, variant="fp16")
# Using channels last layout.
pipe.unet.to(memory_format=torch.channels_last)
pipe.to("cuda") # OR, pipe.enable_sequential_cpu_offload() OR, 

# Generate the base image.
pre_image = base_pipe(prompt=prompt, generator=generator,
        num_inference_steps=total_num_steps, negative_prompt=negative_prompt, num_images_per_prompt=batch_size, output_type="latent" if do_latent else "pil").images

# Generate a range from 0.1 to 0.9, with 0.1 increments.
test_strengths = [0.5]
for refiner_strength in test_strengths:
    # Generate a new set of random states for each image.
    generator_two = torch.Generator(device="cuda").manual_seed(refiner_seed)
    # Put through the refiner now.
    images = pipe(prompt=prompt, image=pre_image, aesthetic_score=10, negative_aesthetic_score=2.4, generator=generator_two,
                num_inference_steps=total_num_steps, num_images_per_prompt=batch_size, strength=refiner_strength, negative_prompt=negative_prompt).images # denoising_start
    for idx in range(0, len(images)):
        print(f'Image: {idx}')
        images[idx].save(f'/notebooks/SimpleTuner/inference/test-{refiner_strength}-{idx}--{batch_size}--{do_latent}.png', format='PNG')
```

**Comment by bghira** [comment]
Children of the Corn with that code

![image](https://github.com/huggingface/diffusers/assets/59658056/980cf5c8-7341-4c5d-923a-679d55c5c8cd)
![image](https://github.com/huggingface/diffusers/assets/59658056/7c083be6-51ed-45b8-8c13-f96b7f032580)

**Comment by patrickvonplaten** [comment]
Thanks checking!

**Comment by patrickvonplaten** [comment]
Just to clarify you see a difference between batch size 1 and batch size 2 here no?

**Comment by bghira** [comment]
checking

**Comment by bghira** [comment]
absolutely. it looks much better.

![image](https://github.com/huggingface/diffusers/assets/59658056/addfdec3-15bd-4f1b-905a-9503e45b556e)

**Comment by bghira** [comment]
seems like the pooled embeds might be getting pooled :dancers:

**Comment by patrickvonplaten** [comment]
Nice I can reproduce it now!

**Comment by bghira** [comment]
YES! You are the best, Patrick!

**Comment by bghira** [comment]
@patrickvonplaten i've updated the title again, i'm super curious what this was. and if there's anything i can do to help. let me know!

**Comment by patrickvonplaten** [comment]
Think I'll find it quickly - I'll make sure to ping you in there PR! You did all the work already by narrowing down the bug to the batch_size > 1 only for img2img :-)

**Comment by bghira** [comment]
i mean, technically we're at a batch size of 1, right? i've always had to set `num_images_per_prompt` to 1, which would be great to fix

**Comment by patrickvonplaten** [comment]
Found it (I think): https://github.com/huggingface/diffusers/pull/4327/files

Could you maybe check with this PR ? :-)

**Comment by patrickvonplaten** [comment]
Can you maybe send your github email (either here or to patrick[at]huggingface.co) so that I can make you a co-author of [the PR](https://github.com/huggingface/diffusers/pull/4327/files) :-) You pretty much did all the work with the rigorous testing here!

**Comment by bghira** [comment]
mine is always the anonymous `bghira@users.github.com` :)

**Comment by bghira** [comment]
Resolved in #4327
